iceoryx2-services/discovery/src/service_discovery/service.rs

```diff
@@ -0,0 +1,331 @@
+    NotifyFailure,
+}
+
+/// Configuration for the service discovery service.
+#[derive(Debug, Clone, serde::Serialize, serde::Deserialize)]
+pub struct Config {
+    /// Whether or not to synchronize the discvery state on initialization.
+    ///
+    /// If enabled, updates for all pre-existing services will not be sent.
+    pub sync_on_initialization: bool,
+
+    /// Whether to include iceoryx-internal services in discovery results.
+    pub include_internal: bool,
+
+    /// Whether to publish discovery events.
+    pub publish_events: bool,
+
+    /// The maximum number of subscribers to the service permitted.
+    pub max_subscribers: usize,
+
+    /// Whether to send notifications on changes.
+    pub send_notifications: bool,
+
+    /// The maximum number of listeners to the service permitted.
+    pub max_listeners: usize,
+}
+
+impl Default for Config {
+    fn default() -> Self {
+        Self {
+            sync_on_initialization: true,
+            include_internal: true,
+            publish_events: true,
+            max_subscribers: 10,
+            send_notifications: true,
+            max_listeners: 10,
+        }
+    }
+}
+
+/// The service discovery service.
+///
+/// This service is responsible for tracking and publishing information about
+/// services in the system. It can detect when services are added or removed,
+/// and notify interested parties about these changes.
+///
+/// # Type Parameters
+///
+/// * `S` - The service type that this discovery service operates on.
+pub struct Service<S: ServiceType> {
+    discovery_config: Config,
+    iceoryx_config: IceoryxConfig,
+    _node: Node<S>,
+    publisher: Option<Publisher<S, [u8], ()>>,
+    notifier: Option<Notifier<S>>,
+    tracker: Tracker<S>,
+}
+
+impl<S: ServiceType> Service<S> {
+    /// Creates the service discovery service.
+    ///
+    /// This function initializes a new service discovery service with the provided
+    /// configuration. The service will track services of type `S` in the system.
+    ///
+    /// # Parameters
+    ///
+    /// * `discovery_config` - Configuration for the discovery service.
+    /// * `iceoryx_config` - Configuration for the underlying iceoryx system.
+    ///
+    /// # Returns
+    ///
+    /// A result containing either the created service or an error if creation failed.
+    ///
+    /// # Errors
+    ///
+    pub fn create(
+        discovery_config: &Config,
+        iceoryx_config: &IceoryxConfig,
+    ) -> Result<Self, CreationError> {
+        let node = NodeBuilder::new()
+            .config(iceoryx_config)
+            .create::<S>()
+            .map_err(|_| CreationError::NodeCreationFailure)?;
+
+        let mut publisher = None;
+        if discovery_config.publish_events {
+            let publish_subscribe = node
+                .service_builder(service_name())
+                .publish_subscribe::<Payload>()
+                .subscriber_max_borrowed_samples(10)
```

Comment:
All of those parameters should also be part of the config.

Reply:
@elfenpiff Added all except for `max_publishers` and `max_notifiers` - there shall only be one.